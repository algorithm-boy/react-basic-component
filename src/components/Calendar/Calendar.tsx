import React, { SFC, ReactElement, Fragment } from 'react';
import { GeneralProps } from '../global.d';
import {
  prevYear,
  prevMonth,
  yesterDay,
  prevHours,
  prevMinutes,
  prevSeconds,
  today,
  nextSeconds,
  nextMinutes,
  nextHours,
  tomorrow,
  nextMonth,
  nextYear,
  DateRange,
  getDays,
  setDate,
} from './date';

export interface CalendarProps extends GeneralProps {
  render: (calendarChange: CalendarChange) => ReactElement;
  startWeekDay?: number;
  rangeCount?: number;
};

export interface CalendarChange {
  prevYear: (date: Date) => Date;
  prevMonth: (date: Date) => Date;
  yesterDay: (date: Date) => Date;
  prevHours: (date: Date) => Date;
  prevMinutes: (date: Date) => Date;
  prevSeconds: (date: Date) => Date;
  today: (date: Date) => Date;
  nextSeconds: (date: Date) => Date;
  nextMinutes: (date: Date) => Date;
  nextHours: (date: Date) => Date;
  tomorrow: (date: Date) => Date;
  nextMonth: (date: Date) => Date;
  nextYear: (date: Date) => Date;
  getDays: (date: Date) => DateRange;
  setDate: (date: Date, n: number, identifier: string) => Date;
}

const Calendar: SFC<CalendarProps> = props => {
  const { render, startWeekDay = 0, rangeCount = 42 } = props;
  const calendarChange = {
    prevYear,
    prevMonth, 
    yesterDay, 
    prevHours, 
    prevMinutes, 
    prevSeconds, 
    today,
    nextSeconds, 
    nextMinutes, 
    nextHours,
    tomorrow,
    nextMonth,
    nextYear,
    setDate,
    getDays: date => getDays(date, startWeekDay, rangeCount),
  } as CalendarChange;
  return render(calendarChange);
};

export default Calendar;
