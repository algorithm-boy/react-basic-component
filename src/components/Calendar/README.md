# Calendar

## API

| 参数(props) | 说明(specification) | 类型(type) | 默认值(defaultProps) |
| --- | --- | --- | --- |
| render | calendar的ui渲染函数 | CalendarChange => React.Element | 必须 |
| startWeekDay | calendar 的起始星期 | number | 0 |
| rangeCount | calendar 的渲染个数 | number | 42 |

## Example

```javascript
<Calendar render={calendarChange => <CalendarUI {...calendarChange} />} />
```

