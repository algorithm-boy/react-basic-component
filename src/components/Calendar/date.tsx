export interface DateRange {
  current: Date;
  range: Date[];
}

export function prevYear(date: Date): Date {
  return new Date(new Date(date).setFullYear(date.getFullYear() - 1));
}

export function prevMonth(date: Date): Date {
  return new Date(new Date(date).setMonth(date.getMonth() - 1));
}

export function yesterDay(date: Date): Date {
  return new Date(new Date(date).setDate(date.getDate() - 1));
}

export function prevHours(date: Date): Date {
  return new Date(new Date(date).setHours(date.getHours() - 1));
}

export function prevMinutes(date: Date): Date {
  return new Date(new Date(date).setMinutes(date.getMinutes() - 1));
}

export function prevSeconds(date: Date): Date {
  return new Date(new Date(date).setSeconds(date.getSeconds() - 1));
}

export function today(date: Date): Date {
  const current = new Date();
  return new Date(current.getFullYear(), current.getMonth(), current.getDate(), date.getHours(), date.getMinutes(), date.getSeconds());
}

export function nextSeconds(date: Date): Date {
  return new Date(new Date(date).setSeconds(date.getSeconds() + 1));
}

export function nextMinutes(date: Date): Date {
  return new Date(new Date(date).setMinutes(date.getMinutes() + 1));
}

export function nextHours(date: Date): Date {
  return new Date(new Date(date).setHours(date.getHours() + 1));
}

export function tomorrow(date: Date): Date {
  return new Date(new Date(date).setDate(date.getDate() + 1));
}

export function nextMonth(date: Date): Date {
  return new Date(new Date(date).setMonth(date.getMonth() + 1));
}

export function nextYear(date: Date): Date {
  return new Date(new Date(date).setFullYear(date.getFullYear() + 1));
}

export function setDate(date: Date, n: number, identifier = 'date') {
  switch(identifier) {
    case 'year': return new Date(new Date(date).setFullYear(n));
    case 'month': return new Date(new Date(date).setMonth(n));
    case 'day': return new Date(new Date(date).setDate(n));
    case 'hour': return new Date(new Date(date).setHours(n));
    case 'minute': return new Date(new Date(date).setMinutes(n));
    case 'second': return new Date(new Date(date).setSeconds(n));
    default: return date;
  } 
}

function isLeapYear(year: number): boolean {
  return !(year % (year % 100 ? 4 : 400));
}

function getMonthDays(date: Date): number {
  const year = date.getFullYear();
  const month = date.getMonth();
  if (isLeapYear(year) && month === 1) {
    return 29;
  } else if (month === 1) {
    return 28;
  } else if ([0, 2, 4, 6, 7, 9, 11].includes(month)) {
    return 31;
  }
  return 30;
}

function getDate(date: Date, day: number): Date {
  return new Date(new Date(date).setDate(day));
}

export function getDays(date: Date, startWeekDay: number, rangeCount: number): DateRange {
  const day = date.getDate();
  const monthDays = getMonthDays(date);
  const leaveRangeCount = rangeCount - 1;
  let leftRangeCount = 0;
  let rightRangeCount = 0;
  const leftRange: Date[] = [];
  const rightRange: Date[] = [];

  if(leaveRangeCount === 0 || rangeCount > 42) {
    return {
      current: date,
      range: [date],
    };
  }

  if(rangeCount >= monthDays) {
    const weekday = date.getDay();
    for(let i = 0; i < rangeCount; i ++) {
      leftRange.push(getDate(date, i + 1));
    }
  }else {
    if(leaveRangeCount % 2 === 0) {
      rightRangeCount = leftRangeCount = leaveRangeCount / 2;
    }else {
      leftRangeCount = (leaveRangeCount - 1) / 2;
      rightRangeCount = (leaveRangeCount + 1) / 2;
    }
  
    for(let i = 0; i < leftRangeCount; i ++) {
      leftRange.unshift(getDate( date, day - i - 1 ));
    }
  
    for(let i = 0; i < rightRangeCount; i ++) {
      rightRange.push(getDate( date, day + i + 1 ));
    }
  }

  let range = rangeCount >= monthDays ? leftRange :  [...leftRange, date, ...rightRange];
  
  const weekday = range[0].getDay();
  let dayCount = 0;

  if(weekday > startWeekDay) {
    dayCount = weekday - startWeekDay;
  }

  if(weekday < startWeekDay) {
    dayCount = (weekday || 7) - startWeekDay;
  }

  if(rangeCount >= 14) {
    if(dayCount == 0) {
      dayCount = 7;
    }
    const firstDate = range[0];
    const day = firstDate.getDate();
    const lastRanges = range.slice(-dayCount).map(x => x.valueOf());
    const supplier = [];
    for(let i = 0; i < dayCount; i ++) {
      supplier.unshift(getDate(firstDate, day - i - 1));
    }

    range = [...supplier, ...range].slice(0, rangeCount);
    if(lastRanges.includes(date.valueOf())) {
      range = range.slice(7);
      const lastDate = range[range.length - 1];
      const lastDay = lastDate.getDate();
      for(let i = 1; i <= 7; i ++) {
        range.push(getDate(lastDate, lastDay + i));
      }
    }
  }

  return {
    current: date,
    range,
  };
}
