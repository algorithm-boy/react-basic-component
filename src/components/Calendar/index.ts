import Calendar from './Calendar';

export { CalendarProps, CalendarChange } from './Calendar';
export default Calendar;
