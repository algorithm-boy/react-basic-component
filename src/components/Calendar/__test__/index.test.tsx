import * as React from 'react';
import { configure, shallow, render, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import Calendar, { CalendarChange } from '../index';

configure({ adapter: new Adapter() });

function CalendarRender(calendarChange: CalendarChange): React.ReactElement {
  class CalendarUI extends React.Component {
    state = {
      years: Array(61).fill(0).map((v, i) => i + 1970),
      months: Array(12).fill(0).map((v, i) => i + 1),
      days: [],
      hours: Array(24).fill(0).map((v, i) => i),
      minutes: Array(60).fill(0).map((v, i) => i),
      seconds: Array(60).fill(0).map((v, i) => i),
      value: new Date(2019, 10, 15),
      range: [] as Date[],
    }

    public componentWillMount() {
      this.selectDate(this.state.value);
    }

    private prevMonth = () => {
      const { value } = this.state;
      const { prevMonth } = calendarChange;
      this.selectDate(prevMonth(value));
    }

    private nextMonth = () => {
      const { value } = this.state;
      const { nextMonth } = calendarChange;
      this.selectDate(nextMonth(value));
    }

    private prevYear = () => {
      const { value } = this.state;
      const { prevYear } = calendarChange;
      this.selectDate(prevYear(value));
    }

    private nextYear = () => {
      const { value } = this.state;
      const { nextYear } = calendarChange;
      this.selectDate(nextYear(value));
    }

    private prevHours = () => {
      const { value } = this.state;
      const { prevHours } = calendarChange;
      this.selectDate(prevHours(value));
    }

    private nextHours = () => {
      const { value } = this.state;
      const { nextHours } = calendarChange;
      this.selectDate(nextHours(value));
    }

    private selectDate = (date: Date) => {
      const { getDays } = calendarChange;
      const { current, range } = getDays(date);
      const daysCount = range.filter(date => date.getMonth() === current.getMonth()).length;
      this.setState({
        value: current,
        days: Array(daysCount).fill(0).map((v, i) => i + 1),
        range,
      });
    }

    private selectYear = (event: React.FormEvent<HTMLSelectElement>) => {
      const { value } = this.state;
      const year = +event.currentTarget.value;
      const { setDate } = calendarChange;
      this.selectDate(setDate(value, year, 'year'));
    }

    private selectMonth = (event: React.FormEvent<HTMLSelectElement>) => {
      const { value } = this.state;
      const month = +event.currentTarget.value;
      const { setDate } = calendarChange;
      this.selectDate(setDate(value, month - 1, 'month'));
    }

    private selectDay = (event: React.FormEvent<HTMLSelectElement>) => {
      const { value } = this.state;
      const day = +event.currentTarget.value;
      const { setDate } = calendarChange;
      this.selectDate(setDate(value, day, 'day'));
    }

    private selectHour = (event: React.FormEvent<HTMLSelectElement>) => {
      const { value } = this.state;
      const hour = +event.currentTarget.value;
      const { setDate } = calendarChange;
      this.selectDate(setDate(value, hour, 'hour'));
    }

    private selectMinute = (event: React.FormEvent<HTMLSelectElement>) => {
      const { value } = this.state;
      const minute = +event.currentTarget.value;
      const { setDate } = calendarChange;
      this.selectDate(setDate(value, minute, 'minute'));
    }

    private selectSecond = (event: React.FormEvent<HTMLSelectElement>) => {
      const { value } = this.state;
      const second = +event.currentTarget.value;
      const { setDate } = calendarChange;
      this.selectDate(setDate(value, second, 'second'));
    }

    public render() {
      const { 
        value, 
        range,
        years,
        months,
        days,
        hours,
        minutes,
        seconds,
      } = this.state;
      const { 
        prevMonth, 
        nextMonth,
        prevYear,
        nextYear,
        prevHours,
        nextHours,
        selectDate,
        selectYear,
        selectMonth,
        selectDay,
        selectHour,
        selectMinute,
        selectSecond,
      } = this;
      return <div>
        <header className="select-header">
          <span>你当前正在选择的日期是: </span>
          <select onChange={selectYear} value={value.getFullYear()}>{ years.map(year => <option value={year} key={year}>{year}</option>) }</select>
          <span> 年 </span>
          <select onChange={selectMonth} value={value.getMonth() + 1}>{ months.map(month => <option value={month} key={month}>{month}</option>) }</select>
          <span> 月 </span>
          <select onChange={selectDay} value={value.getDate()}>{ days.map(day => <option value={day} key={day}>{day}</option>) }</select>
          <span> 日 </span>
          <select onChange={selectHour} value={value.getHours()}>{ hours.map(hour => <option value={hour} key={hour}>{hour}</option>) }</select>
          <span> 时 </span>
          <select onChange={selectMinute} value={value.getMinutes()}>{ minutes.map(minute => <option value={minute} key={minute}>{minute}</option>) }</select>
          <span> 分 </span>
          <select onChange={selectSecond} value={value.getSeconds()}>{ seconds.map(second => <option value={second} key={second}>{second}</option>) }</select>
          <span> 秒 </span>
        </header>
        <article className="date-container">
          {range.map((day, index) => (
            <React.Fragment
              key={ day.valueOf() } 
            >
              <span 
                onClick={() => selectDate(day)} 
                className={`day-cell ${day.valueOf() === value.valueOf() ? 'active-date' : ''} ${day.getMonth() !== value.getMonth() ? 'not-current-month' : ''}`}>{day.getDate().toString().padStart(2, '0')}</span>
              {(index + 1) % 7 == 0 ? <br /> : ''}
            </React.Fragment>
          ))}
        </article>
        <footer>
          <span className="hover-span" onClick={prevMonth}>上月</span>
          <span className="hover-span" onClick={nextMonth}>下月</span>
          <span className="hover-span" onClick={prevYear}>上一年</span>
          <span className="hover-span" onClick={nextYear}>下一年</span>
          <span className="hover-span" onClick={prevHours}>上一小时</span>
          <span className="hover-span" onClick={nextHours}>下一小时</span>
        </footer>
        <style>{`
        .date-container {
          margin: 16px;
        }

        .day-cell{
          display: inline-block;
          width: 30px;
          height: 30px;
          cursor: pointer;
          text-align: center;
          line-height: 30px;
          border: 1px solid #f4f4f4;
          margin-left: -1px;
          margin-top: -1px;
          
        }
        .day-cell:nth-child(7n){
          margin-right: 0;
        }

        .hover-span {
          display: inline-block;
          margin-right: 8px;
          cursor: pointer;
          user-select: none;
          border: 1px solid #1EA7FD;
          padding: 4px 6px;
          border-radius: 8px;
          font-size: 12px;
        }

        .active-date {
          color: #1EA7FD;
        }

        .not-current-month {
          color: #96999f;
        }
      `}</style>
      </div>
    }
  }

  return <CalendarUI />;
}

describe('Calendar', () => {
  it('renders correctly', () => {
    const wrapper = render(
      <Calendar 
        render={CalendarRender} />
    );
    expect(wrapper.html()).toMatchSnapshot(); // 普通的render生成文件很大
  });
});
