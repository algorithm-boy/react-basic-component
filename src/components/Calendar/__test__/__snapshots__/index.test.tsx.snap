// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calendar renders correctly 1`] = `
"<header class=\\"select-header\\"><span>&#x4F60;&#x5F53;&#x524D;&#x6B63;&#x5728;&#x9009;&#x62E9;&#x7684;&#x65E5;&#x671F;&#x662F;: </span><select><option value=\\"1970\\">1970</option><option value=\\"1971\\">1971</option><option value=\\"1972\\">1972</option><option value=\\"1973\\">1973</option><option value=\\"1974\\">1974</option><option value=\\"1975\\">1975</option><option value=\\"1976\\">1976</option><option value=\\"1977\\">1977</option><option value=\\"1978\\">1978</option><option value=\\"1979\\">1979</option><option value=\\"1980\\">1980</option><option value=\\"1981\\">1981</option><option value=\\"1982\\">1982</option><option value=\\"1983\\">1983</option><option value=\\"1984\\">1984</option><option value=\\"1985\\">1985</option><option value=\\"1986\\">1986</option><option value=\\"1987\\">1987</option><option value=\\"1988\\">1988</option><option value=\\"1989\\">1989</option><option value=\\"1990\\">1990</option><option value=\\"1991\\">1991</option><option value=\\"1992\\">1992</option><option value=\\"1993\\">1993</option><option value=\\"1994\\">1994</option><option value=\\"1995\\">1995</option><option value=\\"1996\\">1996</option><option value=\\"1997\\">1997</option><option value=\\"1998\\">1998</option><option value=\\"1999\\">1999</option><option value=\\"2000\\">2000</option><option value=\\"2001\\">2001</option><option value=\\"2002\\">2002</option><option value=\\"2003\\">2003</option><option value=\\"2004\\">2004</option><option value=\\"2005\\">2005</option><option value=\\"2006\\">2006</option><option value=\\"2007\\">2007</option><option value=\\"2008\\">2008</option><option value=\\"2009\\">2009</option><option value=\\"2010\\">2010</option><option value=\\"2011\\">2011</option><option value=\\"2012\\">2012</option><option value=\\"2013\\">2013</option><option value=\\"2014\\">2014</option><option value=\\"2015\\">2015</option><option value=\\"2016\\">2016</option><option value=\\"2017\\">2017</option><option value=\\"2018\\">2018</option><option selected value=\\"2019\\">2019</option><option value=\\"2020\\">2020</option><option value=\\"2021\\">2021</option><option value=\\"2022\\">2022</option><option value=\\"2023\\">2023</option><option value=\\"2024\\">2024</option><option value=\\"2025\\">2025</option><option value=\\"2026\\">2026</option><option value=\\"2027\\">2027</option><option value=\\"2028\\">2028</option><option value=\\"2029\\">2029</option><option value=\\"2030\\">2030</option></select><span> &#x5E74; </span><select><option value=\\"1\\">1</option><option value=\\"2\\">2</option><option value=\\"3\\">3</option><option value=\\"4\\">4</option><option value=\\"5\\">5</option><option value=\\"6\\">6</option><option value=\\"7\\">7</option><option value=\\"8\\">8</option><option value=\\"9\\">9</option><option value=\\"10\\">10</option><option selected value=\\"11\\">11</option><option value=\\"12\\">12</option></select><span> &#x6708; </span><select><option value=\\"1\\">1</option><option value=\\"2\\">2</option><option value=\\"3\\">3</option><option value=\\"4\\">4</option><option value=\\"5\\">5</option><option value=\\"6\\">6</option><option value=\\"7\\">7</option><option value=\\"8\\">8</option><option value=\\"9\\">9</option><option value=\\"10\\">10</option><option value=\\"11\\">11</option><option value=\\"12\\">12</option><option value=\\"13\\">13</option><option value=\\"14\\">14</option><option selected value=\\"15\\">15</option><option value=\\"16\\">16</option><option value=\\"17\\">17</option><option value=\\"18\\">18</option><option value=\\"19\\">19</option><option value=\\"20\\">20</option><option value=\\"21\\">21</option><option value=\\"22\\">22</option><option value=\\"23\\">23</option><option value=\\"24\\">24</option><option value=\\"25\\">25</option><option value=\\"26\\">26</option><option value=\\"27\\">27</option><option value=\\"28\\">28</option><option value=\\"29\\">29</option><option value=\\"30\\">30</option></select><span> &#x65E5; </span><select><option selected value=\\"0\\">0</option><option value=\\"1\\">1</option><option value=\\"2\\">2</option><option value=\\"3\\">3</option><option value=\\"4\\">4</option><option value=\\"5\\">5</option><option value=\\"6\\">6</option><option value=\\"7\\">7</option><option value=\\"8\\">8</option><option value=\\"9\\">9</option><option value=\\"10\\">10</option><option value=\\"11\\">11</option><option value=\\"12\\">12</option><option value=\\"13\\">13</option><option value=\\"14\\">14</option><option value=\\"15\\">15</option><option value=\\"16\\">16</option><option value=\\"17\\">17</option><option value=\\"18\\">18</option><option value=\\"19\\">19</option><option value=\\"20\\">20</option><option value=\\"21\\">21</option><option value=\\"22\\">22</option><option value=\\"23\\">23</option></select><span> &#x65F6; </span><select><option selected value=\\"0\\">0</option><option value=\\"1\\">1</option><option value=\\"2\\">2</option><option value=\\"3\\">3</option><option value=\\"4\\">4</option><option value=\\"5\\">5</option><option value=\\"6\\">6</option><option value=\\"7\\">7</option><option value=\\"8\\">8</option><option value=\\"9\\">9</option><option value=\\"10\\">10</option><option value=\\"11\\">11</option><option value=\\"12\\">12</option><option value=\\"13\\">13</option><option value=\\"14\\">14</option><option value=\\"15\\">15</option><option value=\\"16\\">16</option><option value=\\"17\\">17</option><option value=\\"18\\">18</option><option value=\\"19\\">19</option><option value=\\"20\\">20</option><option value=\\"21\\">21</option><option value=\\"22\\">22</option><option value=\\"23\\">23</option><option value=\\"24\\">24</option><option value=\\"25\\">25</option><option value=\\"26\\">26</option><option value=\\"27\\">27</option><option value=\\"28\\">28</option><option value=\\"29\\">29</option><option value=\\"30\\">30</option><option value=\\"31\\">31</option><option value=\\"32\\">32</option><option value=\\"33\\">33</option><option value=\\"34\\">34</option><option value=\\"35\\">35</option><option value=\\"36\\">36</option><option value=\\"37\\">37</option><option value=\\"38\\">38</option><option value=\\"39\\">39</option><option value=\\"40\\">40</option><option value=\\"41\\">41</option><option value=\\"42\\">42</option><option value=\\"43\\">43</option><option value=\\"44\\">44</option><option value=\\"45\\">45</option><option value=\\"46\\">46</option><option value=\\"47\\">47</option><option value=\\"48\\">48</option><option value=\\"49\\">49</option><option value=\\"50\\">50</option><option value=\\"51\\">51</option><option value=\\"52\\">52</option><option value=\\"53\\">53</option><option value=\\"54\\">54</option><option value=\\"55\\">55</option><option value=\\"56\\">56</option><option value=\\"57\\">57</option><option value=\\"58\\">58</option><option value=\\"59\\">59</option></select><span> &#x5206; </span><select><option selected value=\\"0\\">0</option><option value=\\"1\\">1</option><option value=\\"2\\">2</option><option value=\\"3\\">3</option><option value=\\"4\\">4</option><option value=\\"5\\">5</option><option value=\\"6\\">6</option><option value=\\"7\\">7</option><option value=\\"8\\">8</option><option value=\\"9\\">9</option><option value=\\"10\\">10</option><option value=\\"11\\">11</option><option value=\\"12\\">12</option><option value=\\"13\\">13</option><option value=\\"14\\">14</option><option value=\\"15\\">15</option><option value=\\"16\\">16</option><option value=\\"17\\">17</option><option value=\\"18\\">18</option><option value=\\"19\\">19</option><option value=\\"20\\">20</option><option value=\\"21\\">21</option><option value=\\"22\\">22</option><option value=\\"23\\">23</option><option value=\\"24\\">24</option><option value=\\"25\\">25</option><option value=\\"26\\">26</option><option value=\\"27\\">27</option><option value=\\"28\\">28</option><option value=\\"29\\">29</option><option value=\\"30\\">30</option><option value=\\"31\\">31</option><option value=\\"32\\">32</option><option value=\\"33\\">33</option><option value=\\"34\\">34</option><option value=\\"35\\">35</option><option value=\\"36\\">36</option><option value=\\"37\\">37</option><option value=\\"38\\">38</option><option value=\\"39\\">39</option><option value=\\"40\\">40</option><option value=\\"41\\">41</option><option value=\\"42\\">42</option><option value=\\"43\\">43</option><option value=\\"44\\">44</option><option value=\\"45\\">45</option><option value=\\"46\\">46</option><option value=\\"47\\">47</option><option value=\\"48\\">48</option><option value=\\"49\\">49</option><option value=\\"50\\">50</option><option value=\\"51\\">51</option><option value=\\"52\\">52</option><option value=\\"53\\">53</option><option value=\\"54\\">54</option><option value=\\"55\\">55</option><option value=\\"56\\">56</option><option value=\\"57\\">57</option><option value=\\"58\\">58</option><option value=\\"59\\">59</option></select><span> &#x79D2; </span></header><article class=\\"date-container\\"><span class=\\"day-cell  not-current-month\\">27</span><span class=\\"day-cell  not-current-month\\">28</span><span class=\\"day-cell  not-current-month\\">29</span><span class=\\"day-cell  not-current-month\\">30</span><span class=\\"day-cell  not-current-month\\">31</span><span class=\\"day-cell  \\">01</span><span class=\\"day-cell  \\">02</span><br><span class=\\"day-cell  \\">03</span><span class=\\"day-cell  \\">04</span><span class=\\"day-cell  \\">05</span><span class=\\"day-cell  \\">06</span><span class=\\"day-cell  \\">07</span><span class=\\"day-cell  \\">08</span><span class=\\"day-cell  \\">09</span><br><span class=\\"day-cell  \\">10</span><span class=\\"day-cell  \\">11</span><span class=\\"day-cell  \\">12</span><span class=\\"day-cell  \\">13</span><span class=\\"day-cell  \\">14</span><span class=\\"day-cell active-date \\">15</span><span class=\\"day-cell  \\">16</span><br><span class=\\"day-cell  \\">17</span><span class=\\"day-cell  \\">18</span><span class=\\"day-cell  \\">19</span><span class=\\"day-cell  \\">20</span><span class=\\"day-cell  \\">21</span><span class=\\"day-cell  \\">22</span><span class=\\"day-cell  \\">23</span><br><span class=\\"day-cell  \\">24</span><span class=\\"day-cell  \\">25</span><span class=\\"day-cell  \\">26</span><span class=\\"day-cell  \\">27</span><span class=\\"day-cell  \\">28</span><span class=\\"day-cell  \\">29</span><span class=\\"day-cell  \\">30</span><br><span class=\\"day-cell  not-current-month\\">01</span><span class=\\"day-cell  not-current-month\\">02</span><span class=\\"day-cell  not-current-month\\">03</span><span class=\\"day-cell  not-current-month\\">04</span><span class=\\"day-cell  not-current-month\\">05</span><span class=\\"day-cell  not-current-month\\">06</span><span class=\\"day-cell  not-current-month\\">07</span><br></article><footer><span class=\\"hover-span\\">&#x4E0A;&#x6708;</span><span class=\\"hover-span\\">&#x4E0B;&#x6708;</span><span class=\\"hover-span\\">&#x4E0A;&#x4E00;&#x5E74;</span><span class=\\"hover-span\\">&#x4E0B;&#x4E00;&#x5E74;</span><span class=\\"hover-span\\">&#x4E0A;&#x4E00;&#x5C0F;&#x65F6;</span><span class=\\"hover-span\\">&#x4E0B;&#x4E00;&#x5C0F;&#x65F6;</span></footer><style>
        .date-container {
          margin: 16px;
        }

        .day-cell{
          display: inline-block;
          width: 30px;
          height: 30px;
          cursor: pointer;
          text-align: center;
          line-height: 30px;
          border: 1px solid #f4f4f4;
          margin-left: -1px;
          margin-top: -1px;
          
        }
        .day-cell:nth-child(7n){
          margin-right: 0;
        }

        .hover-span {
          display: inline-block;
          margin-right: 8px;
          cursor: pointer;
          user-select: none;
          border: 1px solid #1EA7FD;
          padding: 4px 6px;
          border-radius: 8px;
          font-size: 12px;
        }

        .active-date {
          color: #1EA7FD;
        }

        .not-current-month {
          color: #96999f;
        }
      </style>"
`;
